File: 프로그래머스/3/42895. N으로 표현/N으로 표현.js
function solution(N, number) {
    const setArr = Array.from({length: 9}, ()=>new Set());
    for (let i=1; i<9; i++){
        setArr[i].add(Number(String(N).repeat(i)));
        if (i>1){
            for (let j=1; j<i; j++){
                for (let x of setArr[j]){
                    for (let y of setArr[i-j]){
                        setArr[i].add(x+y);
                        setArr[i].add(x-y);
                        setArr[i].add(x*y);
                        if (y!==0) setArr[i].add(x/y);
                    }
                }
            }
        }
    }
    for (let i=1; i<9; i++){
        if (setArr[i].has(number)) return i;
    }
    return -1;
}

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.java
import java.util.*;
class Solution {
    public int solution(int N, int number) {
        int answer = 0;
        
        List<Integer>[] canMake = new ArrayList[9];
        for (int i=1; i<9; i++) {
            canMake[i] = new ArrayList<>();
            String five = "";
            for (int j=0; j<i; j++){
                five+=N;
            }
            canMake[i].add(Integer.parseInt(five));
        }
        
        for (int i=2; i<9; i++){
            for (int j=1; j<i; j++) {
                for (Integer num1 : canMake[j]) {
                    for (Integer num2 : canMake[i-j]) {
                        canMake[i].add(num1 + num2);
                        canMake[i].add(num1 - num2);
                        canMake[i].add(num1 * num2);
                        if (num2!=0) canMake[i].add(num1/num2);
                    }
                }
            }
        }
        
        for (int i=1; i<9; i++) {
            if (canMake[i].contains(number)) return i;
        }
        
        return -1;
    }
}